add submitAndWaitBatch sugar function

khancode opened the pull request:
## High Level Overview of Change

Adds a new sugar function `submitAndWaitBatch()` that's an extension of `submitAndWait()` where it takes in an array of transactions with options and asynchronously submits them for each different Sender account.

<!--
Please include a summary/list of the changes.
If too broad, please consider splitting into multiple PRs.
If a relevant Asana task, please link it here.
-->

### Context of Change

Technical Design:
- Similar to `submitAndWait()` but it takes in an array of transactions.
- Each array item has two properties: transaction and opts.
- A promise is created for each different Sender account so that transactions can be submitted asynchronously.
- Therefore, if there are multiple transactions with the same Sender account, then they will be submitted synchronously to ensure their sequence numbers are correctly incremented in each submission.
- TODO: add a flag to check if Mode 1 or 2 is enabled. For now, Mode 1 is enabled by default.
- The mode flag determines what happens after a transaction submission fails for a Sender account:
  - Mode 1: Sender account won't send its remaining transactions and put them in `result.unsubmitted` list.
  - Mode 2: Sender account will continue to submit its remaining transactions.

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

## Test Plan

Added new integ tests.

<!--
Please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

TODO: add a flag to check if Mode 1 or 2 is enabled. For now, Mode 1 is enabled by default.

mvadari commented:
Seems like this doesn't handle errors very elegantly - if one transaction fails, you lose all information about the other transactions.

Also, `autofill` will need to give transactions consecutive sequence numbers, instead of identical ones.

khancode commented:
> Seems like this doesn't handle errors very elegantly - if one transaction fails, you lose all information about the other transactions.
> 
> Also, `autofill` will need to give transactions consecutive sequence numbers, instead of identical ones.

I addressed your comments ✅

justinr1234 commented on packages/xrpl/test/integration/submitAndWaitBatch.test.ts:
```diff
@@ -0,0 +1,481 @@
+        opts: { wallet: this.wallet },
+      },
+      {
+        transaction: invalidAccountDeleteTx,
```
What happens if you submit multiple transactions with a failed one in the middle of the order?

khancode replied:
For a single account, the failed one and subsequent transactions (they will also fail) will be returned in the `result.error` list.

mvadari replied:
The subsequent ones won't fail if you're using tickets, which should be recommended in the docs for this method.

mvadari commented on packages/xrpl/src/sugar/submit.ts:
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable max-depth -- necessary */
```
max-depth seems solvable with helper methods

mvadari commented on packages/xrpl/src/sugar/submit.ts:
```diff
@@ -72,24 +90,77 @@ async function submitAndWait(
+  > = new Map()
+  for (const tx of transactions) {
+    const account = tx.transaction.Account
+    if (!batchMap.has(account)) {
+      batchMap.set(account, [])
+    }
+    batchMap.get(account)?.push(tx)
```
The performance can be improved slightly if this is set up as an if-else

mvadari commented on packages/xrpl/src/sugar/submit.ts:
```diff
@@ -176,6 +247,94 @@ async function waitForFinalTransactionOutcome(
+async function submitAndWaitBatchHelper(
+  client: Client,
+  account: string,
+  batchMap: Map<
```
It would be a little neater if `submitAndWaitBatchHelper` only took `batchMap.get(account)` as a param, instead of the whole map

mvadari commented on packages/xrpl/src/sugar/submit.ts:
```diff
@@ -176,6 +247,94 @@ async function waitForFinalTransactionOutcome(
+    const tx = transactions[idx]
+    try {
+      // eslint-disable-next-line no-await-in-loop -- necessary
+      const txResponse = await submitAndWaitHelper(
```
Most of the advantage of a batch submit and wait method comes from submitting multiple transactions from one account. This structure means that you don't really save any time if you just want to batch a bunch of transactions from one account.

mvadari reviewed (changes requested):
Most of the advantage of a batch submit and wait method comes from submitting multiple transactions from one account, since in prod, most people don't have access to keys for more than one account. This structure means that you don't really save any time if you just want to batch a bunch of transactions from one account, which kind of defeats the purpose of this method.

mvadari commented on packages/xrpl/src/sugar/autofill.ts:
```diff
@@ -57,6 +57,50 @@ async function autofill<T extends Transaction>(
+  transactions: Array<{ transaction: T; signersCount?: number }>,
+): Promise<T[]> {
+  // Sender Account to Sequence map
+  const map: Map<string, number> = new Map()
```
```suggestion
  const accountSequenceMap: Map<string, number> = new Map()
```
I think a descriptive name is actually useful here.

mvadari commented on packages/xrpl/src/sugar/autofill.ts:
```diff
@@ -57,6 +57,50 @@ async function autofill<T extends Transaction>(
+      if (map.has(account)) {
+        transaction.Sequence = map.get(account)
+      } else {
+        // eslint-disable-next-line no-await-in-loop -- necessary
+        await setNextValidSequenceNumber(this, transaction)
```
It doesn't look like you actually set the sequence number here.

khancode replied:
It sets the sequence number inside `setNextValidSequenceNumber` - [source](https://github.com/XRPLF/xrpl.js/blob/76b73e16a97e1a371261b462ee1a24f1c01dbb0c/packages/xrpl/src/sugar/autofill.ts#L140)

mvadari commented on packages/xrpl/src/sugar/autofill.ts:
```diff
@@ -57,6 +57,50 @@ async function autofill<T extends Transaction>(
+      if (map.has(account)) {
+        transaction.Sequence = map.get(account)
+      } else {
+        // eslint-disable-next-line no-await-in-loop -- necessary
```
Seems like you could avoid this by putting everything inside this for loop in a helper method and `Promise.all`ing on that

mvadari commented on packages/xrpl/src/sugar/autofill.ts:
```diff
@@ -88,6 +88,50 @@ async function autofill<T extends Transaction>(
+ */
+async function autofillBatch<T extends Transaction>(
+  this: Client,
+  transactions: Array<{ transaction: T; signersCount?: number }>,
+): Promise<T[]> {
```
Will this typing work with different types of transactions in the same array?

mvadari commented on packages/xrpl/src/sugar/autofill.ts:
```diff
@@ -88,6 +88,50 @@ async function autofill<T extends Transaction>(
+ * Only used for multisigned transactions.
+ * @returns The autofilled transaction.
+ */
+async function autofillBatch<T extends Transaction>(
```
A thought - is it possible to have this be an alternate typing to the standard `autofill` method using typescript generics?

intelliot commented:
notes
- can use Tickets to allow batched transactions to succeed independently
  - Tickets do not improve performance in general
  - they provide some safety in the sense that an invalid transaction won't prevent other transactions from succeeding
- Mayukha indicated this would be useful and she would use it
- valid transactions will succeed. errors should be handled and corrected.

mvadari commented:
> * Mayukha indicated this would be useful and she would use it

I would use it in Python for https://github.com/XRPLF/xbridge-cli

sublimator commented:
> can use Tickets to allow batched transactions to succeed independently
> Tickets do not improve performance in general
> they provide some safety in the sense that an invalid transaction won't prevent other > transactions from succeeding

Way back in the day (2013) ripple-lib had a transaction manager that would handle transaction [re]signing and Sequence'ing. It would even plug Sequence gaps iirc (per Arthur' design) via AccountSet noop transactions (some fun bugs as you might imagine). For various reasons, both compelling and questionable, it was removed in the ill-fated (and some would argue (both then and now) ill-considered) move to RippleAPI. 

It is interesting that 10 years on this is still somewhat of an issue. Having to uses Tickets seems somewhat "expensive".

mvadari commented:
> Having to uses Tickets seems somewhat "expensive".

How so?

sublimator commented:
> Having to uses Tickets seems somewhat "expensive".
> How so?

Well you have to create and delete (consume) Ticket objects in the ledger?
Tickets were in development but canned for a while when I was at ripple, so I don't know if that's something that's changed, but it seems unlikely ?

intelliot commented:
[Tickets](https://xrpl.org/tickets.html) were added by the TicketBatch amendment and enabled on Mainnet in Nov 2021. They add some overhead as objects are created and then consumed, but that's not a particularly high cost compared to many other operations. While I don't think the use of Tickets is necessary for this PR, it would also be perfectly fine to use them if desired.

sublimator commented:
I have nothing against tickets per se, just surprised a bit that txn
submission is still not fully solved

On Fri, 1 Sep 2023 at 11:00 AM Elliot Lee ***@***.***> wrote:

> Tickets <https://xrpl.org/tickets.html> were added by the TicketBatch
> amendment and enabled on Mainnet in Nov 2021. They add some overhead as
> objects are created and then consumed, but that's not a particularly high
> cost compared to many other operations. While I don't think the use of
> Tickets is necessary for this PR, it would also be perfectly fine to use
> them if desired.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/XRPLF/xrpl.js/pull/2149#issuecomment-1702122698>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEAHG7YJ2NG27EPIDORDDLXYFMVRANCNFSM6AAAAAASIFY52M>
> .
> You are receiving this because you commented.Message ID: <XRPLF/xrpl.
> ***@***.***>
>

mvadari commented:
> [Tickets](https://xrpl.org/tickets.html) were added by the TicketBatch amendment and enabled on Mainnet in Nov 2021. They add some overhead as objects are created and then consumed, but that's not a particularly high cost compared to many other operations. While I don't think the use of Tickets is necessary for this PR, it would also be perfectly fine to use them if desired.

I don't think it's worth the complexity of using tickets _in_ the library (you'd need to get a ticket transaction somehow signed by the wallet), the docs for this function should just strongly suggest that the dev use tickets to build the transactions that are being processed by `submitAndWait`.

sublimator commented:
It's tricky. You want to make it easy, but at the same time ...

khancode commented:
Work has been deprioritized, if anyone wants to pick this back up then feel free to copy the changes from this PR.
